contract/RadicleCloudBase.sol

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity >=0.7.0 <0.9.0;
+
+/// @title RadicleCloudBase does on-chain accounting and emits events for operator.
+/// @dev RadicleCloudBase is an abstract contract.
+abstract contract RadicleCloudBase {
+    /// @dev Owner is needed to potentially re-adjust prices in future and withdraw funds.
+    address private owner;
+    /// @notice Keep track of expiry blocks for Orgs
+    mapping(address => uint64) public expiry;
+    /// @notice Remember owner of Org, one who initially created it
+    mapping(address => address) public orgOwner;
+
+    /// @notice Package price ber blockPerMonth
+    uint256 public pricePerMonth;
+    /// @notice Duration of blocks roughly equivalent to one month
+    uint32  public blocksPerMonth;
+
+    /// @notice MAXBLOCK is the maximum value of uint64
+    uint64 private constant MAXBLOCK = type(uint64).max;
+
+    /// @notice Array of when buys happened
+    uint64[] private buys    = [MAXBLOCK];
+    /// @notice Array of when expires happen
+    uint64[] private expires = [MAXBLOCK];
+    /// @notice Array of when cancels happened
+    uint64[] private cancels = [MAXBLOCK];
+    /// @notice Array of when cancel offsets happen
+    uint64[] private offsets = [MAXBLOCK];
+
+    /// @notice Tracks position of last impacted idx for all lasts
+    uint64[] private LAST = [0, 0, 0, 0, 0];
+    uint64 private constant PROCESSED   = 0;
+    uint64 private constant BUY         = 1;
+    uint64 private constant EXPIRE      = 2;
+    uint64 private constant CANCEL      = 3;
+    uint64 private constant OFFSET      = 4;
+
+    /// @notice Rate at which operator currently earns
+    uint64 private RATE = 0;
+
+    struct SubStatus {
+        // block
+        //uint64 previous;
+        // how many blocks have been purchased
+        uint128 shares;
+        // block
+        uint64 next;
+        // rate
+        uint64 consume;
+    }
+
+    /// @notice Subscription status at each block time.
+    mapping(uint64 => SubStatus) public subStatus;
+    /// @dev This is where we start our accounting from.
+    uint64 private lastProcessedSubStatus;
+    /// @dev This is the tip of statuses.
+    uint64 private lastCreatedSubStatus;
+
+    /// @notice Emitted when money is deposited for an Org, could be renewal or initial order.
+    /// @param org The Org being topped up
+    /// @param expiry Next expiry for this Org
+    event NewTopUp(address org, uint64 expiry);
+
+    // modifier to check if caller is owner
+    modifier isOwner() {
+        require(msg.sender == owner, "Caller is not owner");
+        _;
+    }
+
+    /// @dev Set contract deployer as owner.
+    constructor(uint price) {
```

Comment:
We should take the owner as a parameter here, as well as the ERC-20 token address.

Reply:
https://github.com/radicle-dev/radicle-cloud/pull/1/commits/7cd673e5b55b8d1ce2111d81347060fe64c74fd1#diff-6f6505dadea402022048807bd6d0bdfe983d23f83f876025e3eeb17a64ed8d7eR19-R24